### DISCUSSION TITLE
Writing tags in sub-programs

### DISCUSSION THREAD

**Discussion by Rindustrial** [ottowayi/pycomm3#97] [Q&A]
Writing a program tag does not work for me unless I first get the whole tag list.

Am I doing something wrong?

Not a big deal, but it does slow things down a bit. 

This boolean write will not work - 
Ret = plc.write(('Program:M10_AMR.HMI_Inspection_St_Stopper_1_Retract_PB', 0))

... unless I first do this -
Ret = plc.get_tag_list("M10_AMR")

Thanks!

**Comment by Colt-H** [comment]
See issue #40

**Reply by Rindustrial** [reply]
Awesome thanks!

Setting this works for me - 
plc = LogixDriver('192.168.1.10', init_program_tags=True)

**Reply by ottowayi** [reply]
Yep that's the way to do it.  I've went back and forth between whether or not that should be default behavior.  I'm leaning towards making it default, but on the other hand it could add a lot of extra time to the initial upload.  My reasoning is that if you have a lot of program tags, then it's likely you'll want to access them and thus have to spend the time to upload them.  But, if you don't have many program scope tags and don't plan on using them, then there isn't much to upload so won't take much time.  If there are any strong opinions I'm willing to discuss them here and decide which way to go on it.

**Reply by Colt-H** [reply]
In my experience, most integrators I've dealt with either do everything controller scoped, or everything program scoped. I agree that the impact would probably be minimal if you didn't need it on.

Plus if you have the specific case of lots of program tags but only need controller tags, AND need to optimize upload times, you can set it to false.

Plus this is the second issue/discussion opened about it.

**Reply by ottowayi** [reply]
Yeah I agree.  The next version is going to have a ton of internal changes and a couple minor API changes.  Nothing major breaking, but could affect some users.  That would probably be a good time to do it.

**Reply by Colt-H** [reply]
When do you plan to roll it out? If I get time this weekend I can try to update the documentation to explain why it now defaults to true and what changes and use cases for setting to default if you'd like.

I'm not real strong on Python but I can type quickly and want to help if I can.

**Reply by ottowayi** [reply]
I'd love for it to be this weekend, but it depends on how much time I get to work on it.  Most likely it'll be next weekend.

**Reply by ottowayi** [reply]
This was made default starting with the v1.0 releases
